modules/mono/glue/GodotSharp/GodotSharp/Core/NativeInterop/VariantUtils.generic.cs

```diff
@@ -31,380 +33,404 @@ static GenericConversion()
     }
 
     [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
-    public static godot_variant CreateFrom<[MustBeVariant] T>(in T from)
+    public static godot_variant CreateFrom<[MustBeVariant] T>(in T from) => CreateFromLookup<T>.CreateFrom(from);
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static T ConvertTo<[MustBeVariant] T>(in godot_variant variant) => ConvertToLookup<T>.ConvertTo(variant);
+
+    // We use a generic static class to lookup the conversion method once during its initialization.
+    // The .NET runtime does the heavy lifting of creating an instance per type which makes this implementation
+    // behave similar to C++ templates: a specific implementation for each type without any overhead at runtime.
+    private static class ConvertToLookup<[MustBeVariant] T>
+    {
+        private static readonly ConvertToDelegate<T> _converter = DetermineConvertToDelegate<T>();
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
```

Comment:
The usage of `AggressiveOptimization` is actually a deoptimization most of the time as the method skips tiering and cant patriciate in more optimizations that rely on it. The usage of `AggressiveInlining` hides the fact as the caller can still patriciate in tiering and thus allowing it to take advantage of it. So in rare cases when this method is not inlined it ends up generating suboptimal code. Here is simple example demonstrating it:

```csharp
[DisassemblyDiagnoser]
public class InlineBenchmark
{
    [Benchmark]
    public int WithAggressiveOptimization() => Test.WithAggressiveOptimization();

    [Benchmark]
    public int WithoutAggressiveOptimization() => Test.WithoutAggressiveOptimization();
}

public static class Test
{
    private static readonly int value = 0x123456;

    [MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.AggressiveOptimization)]
    public static int WithAggressiveOptimization() => Test.value;

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static int WithoutAggressiveOptimization() => Test.value;
}
```
## .NET 8.0.3 (8.0.324.11423), X64 RyuJIT AVX2
```assembly
; InlineBenchmark.WithAggressiveOptimization()
       jmp       qword ptr [7FF831905ED8]; Test.WithAggressiveOptimization()
; Total bytes of code 6
```
```assembly
; Test.WithAggressiveOptimization()
       sub       rsp,28
       test      byte ptr [7FF8316D1AD6],1
       je        short M01_L01
M01_L00:
       mov       eax,[7FF8316D1B08]
       add       rsp,28
       ret
M01_L01:
       mov       rcx,7FF8316D1AA0
       mov       edx,6
       call      CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
       jmp       short M01_L00
; Total bytes of code 46
```

## .NET 8.0.3 (8.0.324.11423), X64 RyuJIT AVX2
```assembly
; InlineBenchmark.WithoutAggressiveOptimization()
       jmp       qword ptr [7FF831915EF0]; Test.WithoutAggressiveOptimization()
; Total bytes of code 6
```
```assembly
; Test.WithoutAggressiveOptimization()
       mov       eax,123456
       ret
; Total bytes of code 6
```

Reply:
I just copy&pasted the original methods attributes to my newly added functions, it's interesting to see how these can affect the generated code in such a bad way. I wonder what impact removing these all over the C# integration would yield.

I have little experience with the C# JIT, especially on this level. I can remove the attributes but would have to rebenchmark then.

Thank you for the explanation, its always appreciated!

Reply:
Generally you want to avoid using the AO flag as much as possible as the end result is almost always the opposite of what you want. Where it can yield improvement is on a cold start where the JIT compiles methods without optimizations but it also means that the method is never recompiled with better optimizations after the application is on a steady state.

The attribute's documentation mentions that applying the flag without measuring can yield to negative results.

Reply:
I removed `AggressiveOptimization` and gave it a spin locally: the performance seems to be on par for me, a bit faster (14,99ms min time) in one run but that could just be noise. I would have to dig into the generated assembly to see the difference, but I don't know how to do this with Godot and C#.

Reply:
Its generally very hard to benchmark these correctly and the possible difference is just few assembly instructions so its very likely to turn to noise. The expensive part is even inlined so that the only time it jumps to `CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE` is on the first execution.

Also as I mentioned above the `AggressiveInlining` is most likely hiding the negative side-effects of `AggressiveOptimization` in most cases where the method does get inlined as the AO flag is then just ignored. However AI never guarantees that the method is inlined in all cases, so when the method does not get inlined we then get yet another slap in the face by the negative effects of the AO flag.